integration/disk_cleanup_test.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/m3db/m3db/storage"
+
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	errDataCleanupTimedOut = errors.New("cleaning up data files took too long")
+)
+
+func createWriter(storageOpts storage.Options) fs.FileSetWriter {
+	fsOpts := storageOpts.GetCommitLogOptions().GetFilesystemOptions()
+	blockSize := storageOpts.GetRetentionOptions().GetBlockSize()
+	filePathPrefix := fsOpts.GetFilePathPrefix()
+	writerBufferSize := fsOpts.GetWriterBufferSize()
+	newFileMode := fsOpts.GetNewFileMode()
+	newDirectoryMode := fsOpts.GetNewDirectoryMode()
+	return fs.NewWriter(blockSize, filePathPrefix, writerBufferSize, newFileMode, newDirectoryMode)
+}
+
+func createFilesetFiles(t *testing.T, storageOpts storage.Options, shard uint32, fileTimes []time.Time) {
+	writer := createWriter(storageOpts)
+	for _, start := range fileTimes {
+		require.NoError(t, writer.Open(shard, start))
+		require.NoError(t, writer.Close())
+	}
+}
+
+func createCommitLogs(t *testing.T, filePathPrefix string, fileTimes []time.Time) {
+	for _, start := range fileTimes {
+		commitLogFile, _ := fs.NextCommitLogsFile(filePathPrefix, start)
+		_, err := os.Create(commitLogFile)
+		require.NoError(t, err)
+	}
+}
+
+func waitUntilDataCleanedUp(filePathPrefix string, shard uint32, toDelete time.Time, timeout time.Duration) error {
+	dataCleanedUp := func() bool {
+		if fs.FilesetFileExistsAt(filePathPrefix, shard, toDelete) {
+			return false
+		}
+		_, index := fs.NextCommitLogsFile(filePathPrefix, toDelete)
+		if index != 0 {
+			return false
+		}
+		return true
+	}
+	if waitUntil(dataCleanedUp, timeout) {
+		return nil
+	}
+	return errDataCleanupTimedOut
+}
+
+func TestDiskCleanup(t *testing.T) {
+	// Test setup
+	testSetup, err := newTestSetup(newTestOptions())
+	require.NoError(t, err)
+	defer testSetup.close()
+
+	testSetup.storageOpts =
+		testSetup.storageOpts.
+			RetentionOptions(testSetup.storageOpts.GetRetentionOptions().
+				BufferDrain(3 * time.Second).
+				RetentionPeriod(6 * time.Hour))
+
+	blockSize := testSetup.storageOpts.GetRetentionOptions().GetBlockSize()
+	filePathPrefix := testSetup.storageOpts.GetCommitLogOptions().GetFilesystemOptions().GetFilePathPrefix()
+	retentionPeriod := testSetup.storageOpts.GetRetentionOptions().GetRetentionPeriod()
+
+	// Start the server
+	log := testSetup.storageOpts.GetInstrumentOptions().GetLogger()
+	log.Debug("disk cleanup test")
+	require.NoError(t, testSetup.startServer())
+	log.Debug("server is now up")
+
+	// Stop the server
+	defer func() {
+		require.NoError(t, testSetup.stopServer())
+		log.Debug("server is now down")
+	}()
+
+	// Now create some fileset files and commit logs
+	shard := uint32(0)
+	numTimes := 10
+	fileTimes := make([]time.Time, numTimes)
+	now := testSetup.getNowFn()
+	for i := 0; i < numTimes; i++ {
+		fileTimes[i] = now.Add(time.Duration(i) * blockSize)
+	}
+	createFilesetFiles(t, testSetup.storageOpts, shard, fileTimes)
+	createCommitLogs(t, filePathPrefix, fileTimes)
```

Comment:
it's using NextCommitLog API to find the next commit log file to write so it'll just use next index if the server has written some commit logs.